### PR TITLE
7 vote on article

### PR DESCRIPTION
**Pull request by Isaac-Sturtridge**
Adds voting to an article with optimistic rendering. Error handling is a bit basic at the moment, awaiting later taught content & task 12 to focus on it, for now...

### CODE REVIEW

File: src/components/VoteButton.jsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    let inc_votes = hasDownvoted ? 2 : 1
+    if(!hasUpvoted) {
+        setHasUpvoted(true)
+        setHasDownvoted(false)
+        setViewVotes(viewVotes + inc_votes);
+        patchArticle(article_id, inc_votes)
+        .catch((err) => {
+            // a bit difficult to test for this I think but this should update visually if the server goes down
+            // the visual component update in this catch block should be replaced by a real error handler when doing task 12
+            setViewVotes(err)
```
Comment (itskatherine) [inline src/components/VoteButton.jsx]: If the patch request goes wrong (which you can imitate by going to the Network tab in your dev tools  and changing 'no throttling' to 'offline') you set the votes (which starts its life as a number) to a large error object which results in the app crashing. 
You may find this section of the optimistic rendering notes useful in working out how to handle the error here: https://notes.northcoders.com/courses/js-front-end/ui-behaviours-and-optimistic-rendering#error-handling

File: src/components/VoteButton.jsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        });
+    } else {
+        setHasUpvoted(false)
+        setViewVotes(viewVotes - inc_votes)
+        patchArticle(article_id, -inc_votes)
+        .catch((err) => {
+            setViewVotes(err)
+        });
+    }
```
Comment (itskatherine) [inline src/components/VoteButton.jsx:35]: Really nice start when it comes to implementing optimistic rendering here - and you've nearly got it apart from handling when there's an error and the votes need to be rectified to reflect it.

File: src/components/VoteButton.jsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  return (
+    <section className="votingBooth">
+      <button onClick={increaseVotes}>{hasUpvoted ? "Undo upvote" : "Upvote"}</button>
+      {viewVotes}
+      <button onClick={decreaseVotes}>{hasDownvoted ? "Undo downvote" : "Downvote"}</button>
+    </section>
+  );
```
Comment (itskatherine) [inline src/components/VoteButton.jsx:73]: You could consider using something a bit more intuitive like 👍 or 👎 icons to indicate increasing and decreasing votes, or using react-icons
This isn't mandatory, just something you could consider!

**Review by itskatherine — COMMENTED** [review]
Hi Isaac, really solid start just need to fix the error handling of the voting - I've outlined how you can test an error below - once you've done this please resubmit 🙂

**Review by 2202Hannah — CHANGES_REQUESTED** [review]
Hi Isaac, great start implementing Katherine's changes. There is one more thing you need to consider, when the axios request has an error and we go in to the catch block  you will need to make sure that the vote amount reverts back to what it was originally to reflect the patch not working. This will need to be handled in both the increase and decrease votes functions.

File: src/components/VoteButton.jsx (hunk 2 above)
Comment (Harrison365) [reply src/components/VoteButton.jsx:35]: Appears to be fixed :)

File: src/components/VoteButton.jsx (hunk 3 above)
Comment (Harrison365) [reply src/components/VoteButton.jsx:73]: 👍

Comment (Harrison365) [reply src/components/VoteButton.jsx:73]: 👍

File: src/components/VoteButton.jsx (hunk 2 above)
Comment (Harrison365) [reply src/components/VoteButton.jsx:35]: Seems to be fixed 👍

**Pull request merged**